File: Java/Machine Learning/KNN/KNN.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class KNN {

    private int k;
    private List<double[]> X_train;
    private List<Integer> y_train;

    public KNN(int k) {
        this.k = k;
        X_train = new ArrayList<>();
        y_train = new ArrayList<>();
    }

    public void fit(List<double[]> X, List<Integer> y) {
        X_train = X;
        y_train = y;
    }

    public List<Integer> predict(List<double[]> X) {
        List<Integer> y_pred = new ArrayList<>();
        for (double[] x : X) {
            int predictedLabel = _predict(x);
            y_pred.add(predictedLabel);
        }
        return y_pred;
    }

    private int _predict(double[] x) {
        double[] distances = new double[X_train.size()];
        for (int i = 0; i < X_train.size(); i++) {
            distances[i] = euclideanDistance(x, X_train.get(i));
        }

        int[] indices = argsort(distances);
        Map<Integer, Integer> labelCount = new HashMap<>();
        for (int i = 0; i < k; i++) {
            int label = y_train.get(indices[i]);
            labelCount.put(label, labelCount.getOrDefault(label, 0) + 1);
        }

        int mostCommonLabel = -1;
        int maxCount = -1;
        for (Map.Entry<Integer, Integer> entry : labelCount.entrySet()) {
            int label = entry.getKey();
            int count = entry.getValue();
            if (count > maxCount) {
                maxCount = count;
                mostCommonLabel = label;
            }
        }
        return mostCommonLabel;
    }

    private double euclideanDistance(double[] x1, double[] x2) {
        double sum = 0.0;
        for (int i = 0; i < x1.length; i++) {
            sum += Math.pow(x1[i] - x2[i], 2);
        }
        return Math.sqrt(sum);
    }

    private int[] argsort(double[] array) {
        int[] indices = new int[array.length];
        for (int i = 0; i < indices.length; i++) {
            indices[i] = i;
        }
        for (int i = 0; i < indices.length - 1; i++) {
            for (int j = i + 1; j < indices.length; j++) {
                if (array[indices[i]] > array[indices[j]]) {
                    int temp = indices[i];
                    indices[i] = indices[j];
                    indices[j] = temp;
                }
            }
        }
        return indices;
    }

    public static void main(String[] args) {
        List<double[]> X_train = new ArrayList<>();
        X_train.add(new double[] { 1, 2 });
        X_train.add(new double[] { 1.5, 1.8 });
        X_train.add(new double[] { 5, 8 });
        X_train.add(new double[] { 8, 8 });
        X_train.add(new double[] { 1, 0.6 });
        X_train.add(new double[] { 9, 11 });

        List<Integer> y_train = new ArrayList<>();
        y_train.add(0);
        y_train.add(0);
        y_train.add(1);
        y_train.add(1);
        y_train.add(0);
        y_train.add(1);

        KNN knn = new KNN(3);
        knn.fit(X_train, y_train);

        List<double[]> X_test = new ArrayList<>();
        X_test.add(new double[] { 1, 2.5 });
        X_test.add(new double[] { 3, 4 });
        X_test.add(new double[] { 8, 9 });

        List<Integer> y_pred = knn.predict(X_test);

        System.out.println(y_pred);

    }
}

File: Python/Machine Learning/knn.py
'''
Implementation of a KNN algorithm.
Author - Sanket Patil
Requirements:
    - Numpy
Python:
    - 3.8.10
'''

import numpy as np
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split


def euclidean_distance(x1, x2):
    return np.sqrt(np.sum((x1 - x2) ** 2))


class KNN:
    def __init__(self, k=3):
        self.k = k

    def fit(self, X, y):
        self.X_train = X
        self.y_train = y

    def predict(self, X):
        y_pred = [self._predict(x) for x in X]
        return np.array(y_pred)

    # Return the most common class label
    def _predict(self, x):
        distances = [euclidean_distance(x, x_train)
                     for x_train in self.X_train]
        k_indices = np.argsort(distances)[:self.k]
        k_nearest_labels = [self.y_train[i] for i in k_indices]
        most_common = np.argmax(np.bincount(k_nearest_labels))
        return most_common


if __name__ == '__main__':
    dataset = load_breast_cancer()
    X = dataset.data
    y = dataset.target

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=0)

    model = KNN(k=3)
    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)
    cm = confusion_matrix(y_test, y_pred)
    print(cm)
    print(accuracy_score(y_test, y_pred))
